pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloudevent_test.go

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloudevent
+
+import (
+  "encoding/json"
+  "fmt"
+  "regexp"
+	"testing"
+
+  "github.com/google/go-cmp/cmp"
+  "github.com/knative/eventing-sources/pkg/kncloudevents"
+  "github.com/knative/pkg/apis"
+  duckv1beta1 "github.com/knative/pkg/apis/duck/v1beta1"
+  "github.com/tektoncd/pipeline/pkg/logging"
+  "github.com/tektoncd/pipeline/test/names"
+  "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+  corev1 "k8s.io/api/core/v1"
+  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+  defaultSinkURI = "http://sink"
+  invalidSinkURI = "invalid_URI"
+  defaultEventID = "event1234"
+  defaultEventSourceURI = "/taskrun/1234"
+  invalidEventSourceURI = "htt%23p://_invalid_URI##"
+  defaultEventType = TektonTaskRunUnknown
+  taskRunName = "faketaskrunname"
+  invalidConditionSuccessStatus = "foobar"
+)
+
+var (
+  defaultRawData = []byte(`{"metadata": {"name":"faketaskrun"}}`)
+  nilEventType TektonEventType
+  defaultCloudEventClient, _ = kncloudevents.NewDefaultClient()
+  happyClientBehaviour = FakeClientBehaviour{SendSuccessfully: true}
+  failingClientBehaviour = FakeClientBehaviour{SendSuccessfully: false}
+)
+
+func TestSendCloudEvent(t *testing.T) {
+	for _, c := range []struct {
+		desc             string
+    sinkURI          string
+    eventSourceURI   string
+    cloudEventClient CEClient
+    wantErr          bool
+    errRegexp        string
+	}{{
+    desc: "send a cloud event when all inputs are valid",
+    sinkURI: defaultSinkURI,
+    eventSourceURI: defaultEventSourceURI,
+    cloudEventClient: NewFakeClient(&happyClientBehaviour),
+    wantErr: false,
+    errRegexp: "",
+  }, {
+    desc: "send a cloud event with invalid sink URI",
+    sinkURI: invalidSinkURI,
+    eventSourceURI: defaultEventSourceURI,
+    cloudEventClient: defaultCloudEventClient,
+    wantErr: true,
+    errRegexp: fmt.Sprintf("%s: unsupported protocol scheme", invalidSinkURI),
+  }, {
+    desc: "send a cloud event, fail to send",
+    sinkURI: defaultSinkURI,
+    eventSourceURI: defaultEventSourceURI,
+    cloudEventClient: NewFakeClient(&failingClientBehaviour),
+    wantErr: true,
+    errRegexp: fmt.Sprintf("%s had to fail", defaultEventID),
+  }, {
+    desc: "send a cloud event with invalid source URI",
+    sinkURI: defaultSinkURI,
+    eventSourceURI: invalidEventSourceURI,
+    cloudEventClient: defaultCloudEventClient,
+    wantErr: true,
```

Comment:
i find that it can be a lot easier to read and maintain tests if you have a completely different test for the "wantErr" case, i.e. a set of table tests for the success case in one Test function, and another for the error cases. Then you dont need any `if c.wantErr` logic in the tests themselves

Reply:
ack, I'll split the tests.

Reply:
solved in another PR